[FLINK-17161][docs][docker] Document the official docker hub image and examples of how to use it (#12131)
apache/flink: docs/ops/deployment/kubernetes.md

    {% highlight bash %}./bin/flink run -m localhost:8081 ./examples/streaming/WordCount.jar{% endhighlight %}

*  Create a `NodePort` service on the rest service of jobmanager:
    1. Run `kubectl create -f jobmanager-rest-service.yaml` to create the `NodePort` service on jobmanager. The example of `jobmanager-rest-service.yaml` can be found in [appendix](#common-cluster-resource-definitions).
    2. Run `kubectl get svc flink-jobmanager-rest` to know the `node-port` of this service and navigate to [http://&lt;public-node-ip&gt;:&lt;node-port&gt;](http://<public-node-ip>:<node-port>) in your browser.
    3. If you use minikube, you can get its public ip by running `minikube ip`.

>     4. Similarly to `port-forward` solution, you could also use the following command below to submit jobs to the cluster:

Author: I think it is about point 3 of the previous `* kubectl port-forward:` option where you can also use cli but with `localhost` instead of `<public-node-ip>`

        {% highlight bash %}./bin/flink run -m <public-node-ip>:<node-port> ./examples/streaming/WordCount.jar{% endhighlight %}

> In order to terminate the Flink cluster, terminate the specific [Session](#deploy-session-cluster) or [Job](#deploy-job-cluster) cluster components

Reviewer: ```suggestion
In order to terminate the Flink cluster, delete the specific [Session](#deploy-session-cluster) or [Job](#deploy-job-cluster) cluster components
```
To be consistent with the command and the session cluster section; also less repetition which is neat.

and use `kubectl` to terminate the common components:

```sh
    kubectl delete -f jobmanager-service.yaml
    kubectl delete -f flink-configuration-configmap.yaml
    # if created then also the rest service
    kubectl delete -f jobmanager-rest-service.yaml
```

### Deploy Session Cluster

A Flink Session cluster is executed as a long-running Kubernetes Deployment.
Note that you can run multiple Flink jobs on a Session cluster.
Each job needs to be submitted to the cluster after the cluster has been deployed.

> A basic Flink Session cluster deployment in Kubernetes has three components:

Reviewer: What does an advanced Flink Session cluster look like?

Author: I think it implicitly means here that this is just a simple example to start from.

* a Deployment which runs a Flink Master
* a Deployment for a pool of TaskManagers
* a Service exposing the Flink Master's REST and UI ports

After creating [the common cluster components](#deploy-flink-cluster-on-kubernetes), use [the Session specific resource definitions](#session-cluster-resource-definitions)
to launch the Session cluster with the `kubectl` command:

```sh
    kubectl create -f jobmanager-session-deployment.yaml
    kubectl create -f taskmanager-session-deployment.yaml
```

> To terminate the Session cluster, these components can be deleted before [the common ones](#deploy-flink-cluster-on-kubernetes) with the `kubectl` command:

Reviewer: _can_ they be deleted before the common components, or _must_ they be deleted beforehand?

Author: Asaik, in general you create and remove resources unordered in k8s.
I can change this:
`before` -> `along with`

```sh

>     kubectl delete -f taskmanager-session-deployment.yaml

Reviewer: is the order important?

Author: I do not think so

    kubectl delete -f jobmanager-session-deployment.yaml
```

### Deploy Job Cluster

A Flink Job cluster is a dedicated cluster which runs a single job.
The job artifacts should be available locally in the Flink Master container and, thus, there is no extra job submission needed

> (the details are [here](#start-a-single-job-cluster)).

Reviewer: ```suggestion
(You can find more details [here](#start-a-single-job-cluster)).
```

A basic Flink Session cluster deployment in Kubernetes has three components:

* a Job which runs a Flink Master
* a Deployment for a pool of TaskManagers
* a Service exposing the Flink Master's REST and UI ports

Check [the single job specific resource definitions](#job-cluster-resource-definitions) and adjust them accordingly.

The `args` attribute in the `jobmanager-job.yaml` has to specify the main class of the user job.

> Check also [how to specify the Flink Master arguments](docker.html#flink-master-additional-command-line-arguments) to understand

Reviewer: ```suggestion
See also [how to specify the Flink Master arguments](docker.html#flink-master-additional-command-line-arguments) to understand
```
same thing as in the docker docs file.

how to pass other `args` to the Flink image in the `jobmanager-job.yaml`.

The job artifacts should be available from the `job-artifacts-volume` in [the resource definition examples](#job-cluster-resource-definitions).
The definition examples mount the volume as a local directory of the host assuming that you create the components in a minikube cluster.
If you do not use a minikube cluster, you can use any other type of volume, available in your kubernetes cluster, to supply the job artifacts.

> Alternatively, you can build [a custom image](docker.html#start-a-single-job-cluster) which already contains the artifacts then you do not need to mount the volume.

Reviewer: ```suggestion
Alternatively, you can build [a custom image](docker.html#start-a-single-job-cluster) which already contains the artifacts instead.
```

After creating [the common cluster components](#deploy-flink-cluster-on-kubernetes), use [the Job cluster specific resource definitions](#job-cluster-resource-definitions)
to launch the cluster with the `kubectl` command:

```sh
    kubectl create -f jobmanager-job.yaml
    kubectl create -f taskmanager-job-deployment.yaml
```

> To terminate the single job cluster, these components can be deleted before [the common ones](#deploy-flink-cluster-on-kubernetes)

Reviewer: same question as abov, regarding can vs must.

with the `kubectl` command:

```sh

>     kubectl delete -f taskmanager-job-deployment.yaml

Reviewer: same ordering question as above

    blob.server.port: 6124
    jobmanager.rpc.port: 6123
    taskmanager.rpc.port: 6122
    queryable-state.server.ports: 6125

>     jobmanager.memory.process.size: 1600m

Reviewer: Which of these options are truly required, and which are just for demonstration purposes?

Author: I think only `jobmanager.rpc.address` is required

    taskmanager.memory.process.size: 1728m

>     parallelism.default: 2

Reviewer: Bit torn on this one, but IMO we should ensure that the example definitions work properly even if there is only 1 TM.
Right now this could be problematic since parallelism = 2 but numberOfTaskSlots = 1
(so either reduce parallelism or increase slots)

            path: flink-conf.yaml
          - key: log4j.properties
            path: log4j.properties
{% endhighlight %}

### Job cluster resource definitions

`jobmanager-job.yaml`
{% highlight yaml %}

> apiVersion: batch/v1

Reviewer: Purely as a question for me; why are we using 3 different values for `apiVersion`?

Author: I am fully familiar with versioning of Kubernetes components but it matters. The versions correspond to the current examples on Kubernetes docs.

kind: Job
metadata:
  name: flink-jobmanager
spec:
  replicas: 1
  selector:
    matchLabels:
      app: flink
      component: jobmanager
  template:
    metadata:
      labels:
        app: flink
        component: jobmanager
    spec:
      restartPolicy: OnFailure
      containers:
        - name: jobmanager
          image: flink:{% if site.is_stable %}{{site.version}}-scala{{site.scala_version_suffix}}{% else %}latest{% endif %}
          env:

>           workingDir: /opt/flink

Reviewer: is this not controlled by the docker image?

Author: true, this leftover looks redundant
